### PR TITLE
Allow Minors to Change age_range on Source

### PR DESCRIPTION
**Pull request by cassidysymons**

### CODE REVIEW

File: microsetta_interface/implementation.py [hunk 1]
```diff
@@ -79,6 +79,11 @@ class Source:
 ACCT_WRITEABLE_KEYS = [ACCT_FNAME_KEY, ACCT_LNAME_KEY, ACCT_EMAIL_KEY,
                        ACCT_ADDR_KEY, ACCT_LANG_KEY, ACCT_TERMS_KEY]
 
+# Age groups for consent purposes, in order. Order is key as this will be
+# used to govern progression. E.g., a child may move to teen, but a teen may
+# not move to child.
+HUMAN_CONSENT_AGE_GROUPS = ["0-6", "7-12", "13-17", "18-plus"]
```
Comment (wasade) [inline microsetta_interface/implementation.py:85]:
```suggestion
HUMAN_CONSENT_AGE_GROUPS = ("0-6", "7-12", "13-17", "18-plus")
```

File: microsetta_interface/implementation.py [hunk 2]
```diff
@@ -1699,6 +1703,30 @@ def decline_reconsent(*, account_id, source_id):
     )
 
 
+def get_update_age(*, account_id, source_id):
```
Comment (wasade) [inline microsetta_interface/implementation.py:1706]: No prereqs?

File: microsetta_interface/implementation.py [hunk 3]
```diff
@@ -2182,6 +2235,16 @@ def get_consent_view(*, account_id=None, source_id=None, consent_type=None):
...
 
+    consent_type_parts = consent_output['consent_type'].split("_")
+    if consent_type_parts[0] == "adult":
+        age_range = "18-plus"
+    elif consent_type_parts[0] == "adolescent":
+        age_range = "13-17"
+    elif consent_type_parts[0] == "child":
+        age_range = "7-12"
+    else:
+        age_range = "0-6"
```
Comment (wasade) [inline microsetta_interface/implementation.py:2246]: not sure if this is the correct match but essentially let's be defensive and exact

```suggestion
    elif consent_type_lifestage == 'infant':
        age_range = "0-6"
    else:
        raise ValueError(f"This shouldn't happen: {consent_type_lifestage}")
```

File: microsetta_interface/implementation.py [hunk 4]
```diff
@@ -2182,6 +2235,16 @@ def get_consent_view(*, account_id=None, source_id=None, consent_type=None):
         False
     )
 
+    consent_type_parts = consent_output['consent_type'].split("_")
+    if consent_type_parts[0] == "adult":
+        age_range = "18-plus"
+    elif consent_type_parts[0] == "adolescent":
+        age_range = "13-17"
+    elif consent_type_parts[0] == "child":
```
Comment (wasade) [inline microsetta_interface/implementation.py:2243]:
```suggestion
    consent_type_lifestage = consent_output['consent_type'].split("_")[0]
    if consent_type_lifestage == "adult":
        age_range = "18-plus"
    elif consent_type_lifestage == "adolescent":
        age_range = "13-17"
    elif consent_type_lifestage == "child":
```

File: microsetta_interface/implementation.py [hunk 5]
```diff
@@ -3581,6 +3649,10 @@ def check_current_consent(account_id, source_id, consent_type):
     return consent_required["result"]
 
 
+def check_show_update_age(age_range):
+    return age_range != "18-plus"
```
Comment (wasade) [inline microsetta_interface/implementation.py:3653]: We really should look at having these as string literals

File: microsetta_interface/implementation.py [hunk 6]
```diff
@@ -1213,7 +1218,9 @@ def get_consent_page(*, account_id=None):
         participant_name=None,
         age_range=None,
         source_id=None,
-        reconsent=reconsent
+        reconsent=reconsent,
+        cur_age_index=-1,
```
Comment (wasade) [inline microsetta_interface/implementation.py:1222]: We should avoid magic numbers here and in the templates, see comment below

File: microsetta_interface/templates/new_participant.jinja2 [hunk 7]
```diff
@@ -243,18 +258,24 @@
...
-        </div>
-        <div class="col-md">
-            <input type="radio" id="age-range-2" name="age-range" value="7-12" onchange="change_consent_div(this)">
-            <label for="age-range-2" class="consent-age-label">{{ tl['PARTICIPANT_FORM']['AGE_7_12'] |e}}</label>
-        </div>
-        <div class="col-md">
-            <input type="radio" id="age-range-3" name="age-range" value="13-17" onchange="change_consent_div(this)">
-            <label for="age-range-3" class="consent-age-label">{{ tl['PARTICIPANT_FORM']['AGE_13_17'] |e}}</label>
-        </div>
+            {% if cur_age_index <= 0 %}
```
Comment (wasade) [inline microsetta_interface/templates/new_participant.jinja2:261]: This and below is testing against a magic number. Can this be changed to test against an exact value, eg `if cur_age in ('UNKNOWN', '0-6')`? Though ideally these aren't strings but instead string literals though that may be weird crossing the boundary into jinja

**Review by wasade — CHANGES_REQUESTED** [review]

File: microsetta_interface/templates/new_participant.jinja2 (hunk 7 above)
Comment (cassidysymons) [reply microsetta_interface/templates/new_participant.jinja2]: Changed the magic number situation. The age ranges are now string literals in implementation.py, but I agree that copying them into the jinja2 template would be both weird and clunky. Other than that, this PR is ready for re-review, but it won't pass integration tests until the corresponding Private API PR has been merged.

**Comment by wasade** [comment]
Thanks!

**Comment by cassidysymons** [comment]
@wasade Workflow is passing now after merging the corresponding Private API PR. Any further feedback or is this also good to merge?

**Comment by wasade** [comment]
Nah, +1, thanks!

**Pull request merged**